src/main/java/mikera/vectorz/impl/SparseIndexedVector.java

```diff
@@ -145,8 +147,59 @@ public static AVector createFromRow(AMatrix m, int row) {
 	public int nonSparseElementCount() {
 		return data.length;
 	}
-	
-	@Override
+
+    public AVector innerProduct(SparseRowMatrix m) {
+		int cc = m.columnCount();
+		int rc = m.rowCount();
+		checkLength(rc);
+        ASparseIndexedVector r = SparseIndexedVector.createLength(cc);
+		for (int i = 0; i < rc; i++) {
+            AVector row = m.getRow(i);
+            if (row instanceof ZeroVector) {
+                continue;
+            }
+            // TODO: we were casting to ASparseVector, necessary for speed??
+            //            if (row instanceof ASparseVector)
+            // This vector could be of any type, such as Vector3.
+            // And some vectors don't have a sparseClone and instead return
+            // a reference to same instance!!!! (See Vector3...)
+            AVector temp = row.sparseClone();
+            temp.multiply(this.unsafeGet(i));
+            r.add(temp);
+		}
+		return r;
+	}
+
+    public AVector innerProduct(SparseColumnMatrix m) {
+		int cc = m.columnCount();
+		int rc = m.rowCount();
+		checkLength(rc);
+        ASparseIndexedVector r = SparseIndexedVector.createLength(cc);
+		for (int i = 0; i < cc; i++) {
+			r.unsafeSet(i, this.dotProduct((ASparseVector)m.getColumn(i)));
+		}
+		return r;
+	}
+    
+    @Override
+    public AVector innerProduct(AMatrix m) {
+        if (m instanceof SparseRowMatrix) {
```

Comment:
The answer is: yes. `AVector innerProduct(AMatrix)` is the overrode method and other methods with same arity but different types are independent. In other words, Java uses single dispatch for method selection and if our method implementation depends on a more specific type then we must manually handle it.